File: Source/AlphaTab.JavaScript/Importer/Model/JsonConverter.cs
/*
 * This file is part of alphaTab.
 * Copyright © 2018, Daniel Kuschny and Contributors, All rights reserved.
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or at your option any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */

using AlphaTab.Audio.Synth.Midi;
using AlphaTab.Audio.Synth.Midi.Event;
using AlphaTab.Collections;
using AlphaTab.Haxe;
using Haxe.Js.Html;
using Phase;

namespace AlphaTab.Model
{
    /// <summary>
    /// This class can convert a full <see cref="Score"/> instance to a simple JavaScript object and back for further
    /// JSON serialization. 
    /// </summary>
    class JsonConverter
    {
        /// <summary>
        /// Converts the given score into a JSON encoded string. 
        /// </summary>
        /// <param name="score">The score to serialize. </param>
        /// <returns>A JSON encoded string that can be used togehter with <see cref="JSONToScore"/> for conversion.</returns>
        public static string ScoreToJson(Score score)
        {
            var obj = ScoreToJsObject(score);
            return Json.Stringify(obj, (k, v) =>
            {
                if (ArrayBuffer.IsView(v))
                {
                    return Script.Write<object>("untyped __js__(\"Array.apply([], {0})\", v)");
                }
                return v;
            });
        }

        /// <summary>
        /// Converts the given JSON string back to a <see cref="Score"/> object. 
        /// </summary>
        /// <param name="json">The JSON string that was created via <see cref="ScoreToJSON"/></param>
        /// <param name="settings">The settings to use during conversion.</param>
        /// <returns>The converted score object.</returns>
        public static Score JsonToScore(string json, Settings settings = null)
        {
            return JsObjectToScore(JsObjectToScore(Json.Parse(json), settings));
        }

        /// <summary>
        /// Converts the score into a JavaScript object without circular dependencies. 
        /// </summary>
        /// <param name="score">The score object to serialize</param>
        /// <returns>A serialized score object without ciruclar dependencies that can be used for further serializations.</returns>
        public static object ScoreToJsObject(Score score)
        {
            Score score2 = Platform.Platform.NewObject();
            Score.CopyTo(score, score2);
            score2.MasterBars = new FastList<MasterBar>();
            score2.Tracks = new FastList<Track>();

            score2.Stylesheet = Platform.Platform.NewObject();
            RenderStylesheet.CopyTo(score.Stylesheet, score2.Stylesheet);

            #region MasterBars

            for (var i = 0; i < score.MasterBars.Count; i++)
            {
                MasterBar masterBar = score.MasterBars[i];
                MasterBar masterBar2 = Platform.Platform.NewObject();
                MasterBar.CopyTo(masterBar, masterBar2);
                if (masterBar.TempoAutomation != null)
                {
                    masterBar2.TempoAutomation = Platform.Platform.NewObject();
                    Automation.CopyTo(masterBar.TempoAutomation, masterBar2.TempoAutomation);
                }
                if (masterBar.Section != null)
                {
                    masterBar2.Section = Platform.Platform.NewObject();
                    Section.CopyTo(masterBar.Section, masterBar2.Section);
                }

                masterBar2.Fermata = Platform.Platform.NewObject();
                foreach (var offset in masterBar.Fermata)
                {
                    var fermata = masterBar.Fermata[offset];
                    var fermata2 = masterBar2.Fermata[offset] = Platform.Platform.NewObject();
                    Fermata.CopyTo(fermata, fermata2);
                }

                score2.MasterBars.Add(masterBar2);
            }

            #endregion

            #region Tracks

            for (int t = 0; t < score.Tracks.Count; t++)
            {
                var track = score.Tracks[t];
                Track track2 = Platform.Platform.NewObject();
                track2.Color = Platform.Platform.NewObject();
                Track.CopyTo(track, track2);

                track2.PlaybackInfo = Platform.Platform.NewObject();
                PlaybackInformation.CopyTo(track.PlaybackInfo, track2.PlaybackInfo);


                #region Staves
                track2.Staves = new FastList<Staff>();

                for (int s = 0; s < track.Staves.Count; s++)
                {
                    var staff = track.Staves[s];
                    Staff staff2 = Platform.Platform.NewObject();
                    Staff.CopyTo(staff, staff2);

                    staff2.Chords = new FastDictionary<string, Chord>();
                    foreach (var key in staff.Chords)
                    {
                        var chord = staff.Chords[key];
                        Chord chord2 = Platform.Platform.NewObject();
                        Chord.CopyTo(chord, chord2);
                        staff2.Chords[key] = chord;
                    }

                    #region Bars

                    staff2.Bars = new FastList<Bar>();
                    for (int b = 0; b < staff.Bars.Count; b++)
                    {
                        var bar = staff.Bars[b];
                        Bar bar2 = Platform.Platform.NewObject();
                        Bar.CopyTo(bar, bar2);

                        #region Voices

                        bar2.Voices = new FastList<Voice>();
                        for (int v = 0; v < bar.Voices.Count; v++)
                        {
                            var voice = bar.Voices[v];
                            Voice voice2 = Platform.Platform.NewObject();
                            Voice.CopyTo(voice, voice2);

                            #region Beats

                            voice2.Beats = new FastList<Beat>();
                            for (int bb = 0; bb < voice.Beats.Count; bb++)
                            {
                                var beat = voice.Beats[bb];
                                Beat beat2 = Platform.Platform.NewObject();
                                Beat.CopyTo(beat, beat2);

                                beat2.Automations = new FastList<Automation>();
                                for (int a = 0; a < beat.Automations.Count; a++)
                                {
                                    Automation automation = Platform.Platform.NewObject();
                                    Automation.CopyTo(beat.Automations[a], automation);
                                    beat2.Automations.Add(automation);
                                }

                                beat2.WhammyBarPoints = new FastList<BendPoint>();
                                for (int i = 0; i < beat.WhammyBarPoints.Count; i++)
                                {
                                    BendPoint point = Platform.Platform.NewObject();
                                    BendPoint.CopyTo(beat.WhammyBarPoints[i], point);
                                    beat2.WhammyBarPoints.Add(point);
                                }

                                #region Notes

                                beat2.Notes = new FastList<Note>();
                                for (int n = 0; n < beat.Notes.Count; n++)
                                {
                                    var note = beat.Notes[n];
                                    Note note2 = Platform.Platform.NewObject();
                                    Note.CopyTo(note, note2);

                                    note2.BendPoints = new FastList<BendPoint>();
                                    for (int i = 0; i < note.BendPoints.Count; i++)
                                    {
                                        BendPoint point = Platform.Platform.NewObject();
                                        BendPoint.CopyTo(note.BendPoints[i], point);
                                        note2.BendPoints.Add(point);
                                    }

                                    beat2.Notes.Add(note2);
                                }

                                #endregion

                                voice2.Beats.Add(beat2);
                            }

                            #endregion

                            bar2.Voices.Add(voice2);
                        }

                        #endregion

                        staff2.Bars.Add(bar2);
                    }

                    #endregion
                    track2.Staves.Add(staff2);
                }

                #endregion

                score2.Tracks.Add(track2);
            }

            #endregion

            return score2;
        }

        /// <summary>
        /// Converts the given JavaScript object into a score object. 
        /// </summary>
        /// <param name="jsObject">The javascript object created via <see cref="ScoreToJsObject"/></param>
        /// <param name="settings">The settings to use during conversion.</param>
        /// <returns>The converted score object.</returns>
        public static Score JsObjectToScore(object jsObject, Settings settings = null)
        {
            Score score = jsObject.As<Score>();
            var score2 = new Score();
            Score.CopyTo(score, score2);
            RenderStylesheet.CopyTo(score.Stylesheet, score2.Stylesheet);

            #region MasterBars

            for (var i = 0; i < score.MasterBars.Count; i++)
            {
                var masterBar = score.MasterBars[i];
                var masterBar2 = new MasterBar();
                MasterBar.CopyTo(masterBar, masterBar2);
                if (masterBar.TempoAutomation != null)
                {
                    masterBar2.TempoAutomation = new Automation();
                    Automation.CopyTo(masterBar.TempoAutomation, masterBar2.TempoAutomation);
                }
                if (masterBar.Section != null)
                {
                    masterBar2.Section = new Section();
                    Section.CopyTo(masterBar.Section, masterBar2.Section);
                }

                foreach (var offset in masterBar.Fermata)
                {
                    var fermata = masterBar.Fermata[offset];
                    var fermata2 = new Fermata();
                    Fermata.CopyTo(fermata, fermata2);
                    masterBar2.AddFermata(offset, fermata2);
                }


                score2.AddMasterBar(masterBar2);
            }

            #endregion

            #region Tracks

            for (int t = 0; t < score.Tracks.Count; t++)
            {
                var track = score.Tracks[t];
                var track2 = new Track(track.Staves.Count);
                Track.CopyTo(track, track2);
                score2.AddTrack(track2);

                PlaybackInformation.CopyTo(track.PlaybackInfo, track2.PlaybackInfo);


                #region Staves

                for (var s = 0; s < track.Staves.Count; s++)
                {
                    var staff = track.Staves[s];
                    var staff2 = track2.Staves[s];
                    Staff.CopyTo(staff, staff2);


                    foreach (var key in staff.Chords)
                    {
                        var chord = staff.Chords[key];
                        var chord2 = new Chord();
                        Chord.CopyTo(chord, chord2);
                        staff2.Chords[key] = chord2;
                    }
                    #region Bars

                    for (int b = 0; b < staff.Bars.Count; b++)
                    {
                        var bar = staff.Bars[b];
                        var bar2 = new Bar();
                        Bar.CopyTo(bar, bar2);
                        staff2.AddBar(bar2);

                        #region Voices

                        for (int v = 0; v < bar.Voices.Count; v++)
                        {
                            var voice = bar.Voices[v];
                            var voice2 = new Voice();
                            Voice.CopyTo(voice, voice2);
                            bar2.AddVoice(voice2);

                            #region Beats

                            for (int bb = 0; bb < voice.Beats.Count; bb++)
                            {
                                var beat = voice.Beats[bb];
                                var beat2 = new Beat();
                                Beat.CopyTo(beat, beat2);
                                voice2.AddBeat(beat2);

                                for (int a = 0; a < beat.Automations.Count; a++)
                                {
                                    var automation = new Automation();
                                    Automation.CopyTo(beat.Automations[a], automation);
                                    beat2.Automations.Add(automation);
                                }

                                for (int i = 0; i < beat.WhammyBarPoints.Count; i++)
                                {
                                    var point = new BendPoint();
                                    BendPoint.CopyTo(beat.WhammyBarPoints[i], point);
                                    beat2.AddWhammyBarPoint(point);
                                }

                                #region Notes

                                for (int n = 0; n < beat.Notes.Count; n++)
                                {
                                    var note = beat.Notes[n];
                                    var note2 = new Note();
                                    Note.CopyTo(note, note2);
                                    beat2.AddNote(note2);

                                    for (int i = 0; i < note.BendPoints.Count; i++)
                                    {
                                        var point = new BendPoint();
                                        BendPoint.CopyTo(note.BendPoints[i], point);
                                        note2.AddBendPoint(point);
                                    }
                                }

                                #endregion
                            }

                            #endregion
                        }

                        #endregion
                    }

                    #endregion

                }
                #endregion
            }

            #endregion

            score2.Finish(settings);
            return score2;
        }

        internal static MidiFile JsObjectToMidiFile(dynamic midi)
        {
            var midi2 = new MidiFile();
            midi2.Division = midi.Division;

            FastList<dynamic> midiEvents = midi.Events;
            foreach (var midiEvent in midiEvents)
            {
                int tick = midiEvent.Tick;
                int message = midiEvent.Message;
                MidiEvent midiEvent2;
                switch (midiEvent.Type)
                {
                    case "alphaTab.audio.synth.midi.event.SystemExclusiveEvent":
                        midiEvent2 = new SystemExclusiveEvent(tick, 0, 0, midiEvent.Data);
                        midiEvent2.Message = message;
                        break;
                    case "alphaTab.audio.synth.midi.event.MetaDataEvent":
                        midiEvent2 = new MetaDataEvent(tick, 0, 0, midiEvent.Data);
                        midiEvent2.Message = message;
                        break;
                    case "alphaTab.audio.synth.midi.event.MetaNumberEvent":
                        midiEvent2 = new MetaNumberEvent(tick, 0, 0, midiEvent.Value);
                        midiEvent2.Message = message;
                        break;
                    default:
                        midiEvent2 = new MidiEvent(tick, 0, 0, 0);
                        midiEvent2.Message = message;
                        break;
                }
                midi2.Events.Add(midiEvent2);
            }

            return midi2;
        }

        internal static object MidiFileToJsObject(MidiFile midi)
        {
            var midi2 = Platform.Platform.NewObject();
            midi2.Division = midi.Division;
            var midiEvents = new FastList<object>();
            midi2.Events = midiEvents;
            foreach (var midiEvent in midi.Events)
            {
                var midiEvent2 = Platform.Platform.NewObject();
                midiEvents.Add(midiEvent2);
                midiEvent2.Type = Platform.Platform.GetTypeName(midiEvent);
                midiEvent2.Tick = midiEvent.Tick;
                midiEvent2.Message = midiEvent.Message;
                switch (midiEvent2.Type)
                {
                    case "alphaTab.audio.synth.midi.event.SystemExclusiveEvent":
                        SystemExclusiveEvent sysex = (SystemExclusiveEvent)midiEvent;
                        midiEvent2.Data = sysex.Data;
                        break;
                    case "alphaTab.audio.synth.midi.event.MetaDataEvent":
                        MetaDataEvent metadata = (MetaDataEvent)midiEvent;
                        midiEvent2.Data = metadata.Data;
                        break;
                    case "alphaTab.audio.synth.midi.event.MetaNumberEvent":
                        MetaNumberEvent metanumber = (MetaNumberEvent)midiEvent;
                        midiEvent2.Value = metanumber.Value;
                        break;
                }
            }

            return midi2;
        }
    }
}

File: src/model/JsonConverter.ts
import {
    AlphaTabMetronomeEvent,
    AlphaTabRestEvent,
    ControlChangeEvent,
    EndOfTrackEvent,
    type MidiEvent,
    MidiEventType,
    NoteBendEvent,
    type NoteEvent,
    NoteOffEvent,
    NoteOnEvent,
    PitchBendEvent,
    ProgramChangeEvent,
    TempoChangeEvent,
    TimeSignatureEvent
} from '@src/midi/MidiEvent';
import { MidiFile, MidiTrack } from '@src/midi/MidiFile';
import { Score } from '@src/model/Score';
import { Settings } from '@src/Settings';
import { ScoreSerializer } from '@src/generated/model/ScoreSerializer';
import { SettingsSerializer } from '@src/generated/SettingsSerializer';
import { AlphaTabError, AlphaTabErrorType } from '@src/AlphaTabError';
import type { ControllerType } from '@src/midi/ControllerType';
import { JsonHelper } from '@src/io/JsonHelper';

/**
 * This class can convert a full {@link Score} instance to a simple JavaScript object and back for further
 * JSON serialization.
 */
export class JsonConverter {
    /**
     * @target web
     */
    private static jsonReplacer(_: any, v: any) {
        if (v instanceof Map) {
            if ('fromEntries' in Object) {
                return Object.fromEntries(v);
            }
            const o: any = {};
            for (const [k, mv] of v) {
                o[k] = mv;
            }
            return o;
        }
        if (ArrayBuffer.isView(v)) {
            return Array.apply([], [v]);
        }
        return v;
    }

    /**
     * Converts the given score into a JSON encoded string.
     * @param score The score to serialize.
     * @returns A JSON encoded string.
     * @target web
     */
    public static scoreToJson(score: Score): string {
        const obj: unknown = JsonConverter.scoreToJsObject(score);
        return JSON.stringify(obj, JsonConverter.jsonReplacer);
    }

    /**
     * Converts the given JSON string back to a {@link Score} object.
     * @param json The JSON string
     * @param settings The settings to use during conversion.
     * @returns The converted score object.
     * @target web
     */
    public static jsonToScore(json: string, settings?: Settings): Score {
        return JsonConverter.jsObjectToScore(JSON.parse(json), settings);
    }

    /**
     * Converts the score into a JavaScript object without circular dependencies.
     * @param score The score object to serialize
     * @returns A serialized score object without ciruclar dependencies that can be used for further serializations.
     */
    public static scoreToJsObject(score: Score): unknown {
        return ScoreSerializer.toJson(score);
    }

    /**
     * Converts the given JavaScript object into a score object.
     * @param jsObject The javascript object created via {@link Score}
     * @param settings The settings to use during conversion.
     * @returns The converted score object.
     */
    public static jsObjectToScore(jsObject: unknown, settings?: Settings): Score {
        const score: Score = new Score();
        ScoreSerializer.fromJson(score, jsObject);
        score.finish(settings ?? new Settings());
        return score;
    }

    /**
     * Converts the given settings into a JSON encoded string.
     * @param settings The settings to serialize.
     * @returns A JSON encoded string.
     * @target web
     */
    public static settingsToJson(settings: Settings): string {
        const obj: unknown = JsonConverter.settingsToJsObject(settings);
        return JSON.stringify(obj, JsonConverter.jsonReplacer);
    }

    /**
     * Converts the given JSON string back to a {@link Score} object.
     * @param json The JSON string
     * @returns The converted settings object.
     * @target web
     */
    public static jsonToSettings(json: string): Settings {
        return JsonConverter.jsObjectToSettings(JSON.parse(json));
    }

    /**
     * Converts the settings object into a JavaScript object for transmission between components or saving purposes.
     * @param settings The settings object to serialize
     * @returns A serialized settings object without ciruclar dependencies that can be used for further serializations.
     */
    public static settingsToJsObject(settings: Settings): Map<string, unknown> | null {
        return SettingsSerializer.toJson(settings);
    }

    /**
     * Converts the given JavaScript object into a settings object.
     * @param jsObject The javascript object created via {@link Settings}
     * @returns The converted Settings object.
     */
    public static jsObjectToSettings(jsObject: unknown): Settings {
        const settings: Settings = new Settings();
        SettingsSerializer.fromJson(settings, jsObject);
        return settings;
    }

    /**
     * Converts the given JavaScript object into a MidiFile object.
     * @param jsObject The javascript object to deserialize.
     * @returns The converted MidiFile.
     */
    public static jsObjectToMidiFile(jsObject: unknown): MidiFile {
        const midi2: MidiFile = new MidiFile();

        JsonHelper.forEach(jsObject, (v, k) => {
            switch (k) {
                case 'division':
                    midi2.division = v as number;
                    break;
                case 'tracks':
                    for (const midiTrack of v as unknown[]) {
                        const midiTrack2: MidiTrack = JsonConverter.jsObjectToMidiTrack(midiTrack);
                        midi2.tracks.push(midiTrack2);
                    }
                    break;
            }
        });

        return midi2;
    }

    private static jsObjectToMidiTrack(jsObject: unknown): MidiTrack {
        const midi2: MidiTrack = new MidiTrack();

        JsonHelper.forEach(jsObject, (v, k) => {
            switch (k) {
                case 'events':
                    for (const midiEvent of v as unknown[]) {
                        const midiEvent2: MidiEvent = JsonConverter.jsObjectToMidiEvent(midiEvent);
                        midi2.events.push(midiEvent2);
                    }
                    break;
            }
        });

        return midi2;
    }

    /**
     * Converts the given JavaScript object into a MidiEvent object.
     * @param jsObject The javascript object to deserialize.
     * @returns The converted MidiEvent.
     */
    public static jsObjectToMidiEvent(midiEvent: unknown): MidiEvent {
        const track: number = JsonHelper.getValue(midiEvent, 'track') as number;
        const tick: number = JsonHelper.getValue(midiEvent, 'tick') as number;
        const type: MidiEventType = JsonHelper.getValue(midiEvent, 'type') as number as MidiEventType;

        switch (type) {
            case MidiEventType.TimeSignature:
                return new TimeSignatureEvent(
                    track,
                    tick,
                    JsonHelper.getValue(midiEvent, 'numerator') as number,
                    JsonHelper.getValue(midiEvent, 'denominatorIndex') as number,
                    JsonHelper.getValue(midiEvent, 'midiClocksPerMetronomeClick') as number,
                    JsonHelper.getValue(midiEvent, 'thirdySecondNodesInQuarter') as number
                );
            case MidiEventType.AlphaTabRest:
                return new AlphaTabRestEvent(track, tick, JsonHelper.getValue(midiEvent, 'channel') as number);
            case MidiEventType.AlphaTabMetronome:
                return new AlphaTabMetronomeEvent(
                    track,
                    tick,
                    JsonHelper.getValue(midiEvent, 'metronomeNumerator') as number,
                    JsonHelper.getValue(midiEvent, 'metronomeDurationInTicks') as number,
                    JsonHelper.getValue(midiEvent, 'metronomeDurationInMilliseconds') as number
                );
            case MidiEventType.NoteOn:
                return new NoteOnEvent(
                    track,
                    tick,
                    JsonHelper.getValue(midiEvent, 'channel') as number,
                    JsonHelper.getValue(midiEvent, 'noteKey') as number,
                    JsonHelper.getValue(midiEvent, 'noteVelocity') as number
                );
            case MidiEventType.NoteOff:
                return new NoteOffEvent(
                    track,
                    tick,
                    JsonHelper.getValue(midiEvent, 'channel') as number,
                    JsonHelper.getValue(midiEvent, 'noteKey') as number,
                    JsonHelper.getValue(midiEvent, 'noteVelocity') as number
                );
            case MidiEventType.ControlChange:
                return new ControlChangeEvent(
                    track,
                    tick,
                    JsonHelper.getValue(midiEvent, 'channel') as number,
                    JsonHelper.getValue(midiEvent, 'controller') as number as ControllerType,
                    JsonHelper.getValue(midiEvent, 'value') as number
                );
            case MidiEventType.ProgramChange:
                return new ProgramChangeEvent(
                    track,
                    tick,
                    JsonHelper.getValue(midiEvent, 'channel') as number,
                    JsonHelper.getValue(midiEvent, 'program') as number
                );
            case MidiEventType.TempoChange:
                const tempo = new TempoChangeEvent(tick, 0);
                tempo.beatsPerMinute = JsonHelper.getValue(midiEvent, 'beatsPerMinute') as number;
                return tempo;
            case MidiEventType.PitchBend:
                return new PitchBendEvent(
                    track,
                    tick,
                    JsonHelper.getValue(midiEvent, 'channel') as number,
                    JsonHelper.getValue(midiEvent, 'value') as number
                );
            case MidiEventType.PerNotePitchBend:
                return new NoteBendEvent(
                    track,
                    tick,
                    JsonHelper.getValue(midiEvent, 'channel') as number,
                    JsonHelper.getValue(midiEvent, 'noteKey') as number,
                    JsonHelper.getValue(midiEvent, 'value') as number
                );
            case MidiEventType.EndOfTrack:
                return new EndOfTrackEvent(track, tick);
        }

        throw new AlphaTabError(AlphaTabErrorType.Format, `Unknown Midi Event type: ${type}`);
    }

    /**
     * Converts the given MidiFile object into a serialized JavaScript object.
     * @param midi The midi file to convert.
     * @returns A serialized MidiFile object without ciruclar dependencies that can be used for further serializations.
     */
    public static midiFileToJsObject(midi: MidiFile): Map<string, unknown> {
        const o = new Map<string, unknown>();
        o.set('division', midi.division);

        const tracks: Map<string, unknown>[] = [];
        for (const track of midi.tracks) {
            tracks.push(JsonConverter.midiTrackToJsObject(track));
        }
        o.set('tracks', tracks);

        return o;
    }

    private static midiTrackToJsObject(midi: MidiTrack): Map<string, unknown> {
        const o = new Map<string, unknown>();

        const events: Map<string, unknown>[] = [];
        for (const track of midi.events) {
            events.push(JsonConverter.midiEventToJsObject(track));
        }
        o.set('events', events);

        return o;
    }

    /**
     * Converts the given MidiEvent object into a serialized JavaScript object.
     * @param midi The midi file to convert.
     * @returns A serialized MidiEvent object without ciruclar dependencies that can be used for further serializations.
     */
    public static midiEventToJsObject(midiEvent: MidiEvent): Map<string, unknown> {
        const o = new Map<string, unknown>();
        o.set('track', midiEvent.track);
        o.set('tick', midiEvent.tick);
        o.set('type', midiEvent.type as number);
        switch (midiEvent.type) {
            case MidiEventType.TimeSignature:
                o.set('numerator', (midiEvent as TimeSignatureEvent).numerator);
                o.set('denominatorIndex', (midiEvent as TimeSignatureEvent).denominatorIndex);
                o.set('midiClocksPerMetronomeClick', (midiEvent as TimeSignatureEvent).midiClocksPerMetronomeClick);
                o.set('thirdySecondNodesInQuarter', (midiEvent as TimeSignatureEvent).thirtySecondNodesInQuarter);
                break;
            case MidiEventType.AlphaTabRest:
                o.set('channel', (midiEvent as AlphaTabRestEvent).channel);
                break;
            case MidiEventType.AlphaTabMetronome:
                o.set('metronomeNumerator', (midiEvent as AlphaTabMetronomeEvent).metronomeNumerator);
                o.set(
                    'metronomeDurationInMilliseconds',
                    (midiEvent as AlphaTabMetronomeEvent).metronomeDurationInMilliseconds
                );
                o.set('metronomeDurationInTicks', (midiEvent as AlphaTabMetronomeEvent).metronomeDurationInTicks);
                break;
            case MidiEventType.NoteOn:
            case MidiEventType.NoteOff:
                o.set('channel', (midiEvent as NoteEvent).channel);
                o.set('noteKey', (midiEvent as NoteEvent).noteKey);
                o.set('noteVelocity', (midiEvent as NoteEvent).noteVelocity);
                break;
            case MidiEventType.ControlChange:
                o.set('channel', (midiEvent as ControlChangeEvent).channel);
                o.set('controller', (midiEvent as ControlChangeEvent).controller as number);
                o.set('value', (midiEvent as ControlChangeEvent).value);
                break;
            case MidiEventType.ProgramChange:
                o.set('channel', (midiEvent as ProgramChangeEvent).channel);
                o.set('program', (midiEvent as ProgramChangeEvent).program);
                break;
            case MidiEventType.TempoChange:
                o.set('beatsPerMinute', (midiEvent as TempoChangeEvent).beatsPerMinute);
                break;
            case MidiEventType.PitchBend:
                o.set('channel', (midiEvent as PitchBendEvent).channel);
                o.set('value', (midiEvent as PitchBendEvent).value);
                break;
            case MidiEventType.PerNotePitchBend:
                o.set('channel', (midiEvent as NoteBendEvent).channel);
                o.set('noteKey', (midiEvent as NoteBendEvent).noteKey);
                o.set('value', (midiEvent as NoteBendEvent).value);
                break;
            case MidiEventType.EndOfTrack:
                break;
        }

        return o;
    }
}
